controller/addon.go

```diff
@@ -0,0 +1,163 @@
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/client-go/tools/record"
+	k8s "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	Name = "addon"
+)
+
+//
+// Package logger.
+var log = logging.WithName(Name)
+
+//
+// Settings defines applcation settings.
+var Settings = &settings.Settings
+
+//
+// Add the controller.
+func Add(mgr manager.Manager, db *gorm.DB) error {
+	reconciler := &Reconciler{
+		EventRecorder: mgr.GetRecorder(Name),
+		Client:        mgr.GetClient(),
+		Log:           log,
+		DB:            db,
+	}
+	cnt, err := controller.New(
+		Name,
+		mgr,
+		controller.Options{
+			Reconciler: reconciler,
+		})
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+	// Primary CR.
+	err = cnt.Watch(
+		&source.Kind{Type: &api.Addon{}},
+		&handler.EnqueueRequestForObject{})
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	return nil
+}
+
+//
+// Reconciler reconciles addon CRs.
+type Reconciler struct {
+	record.EventRecorder
+	k8s.Client
+	DB  *gorm.DB
+	Log *logging.Logger
+}
+
+//
+// Reconcile a Addon CR.
+// Note: Must not a pointer receiver to ensure that the
+// logger and other state is not shared.
+func (r Reconciler) Reconcile(request reconcile.Request) (result reconcile.Result, err error) {
+	r.Log = logging.WithName(
+		names.SimpleNameGenerator.GenerateName(Name+"|"),
+		"addon",
+		request)
+
+	// Fetch the CR.
+	addon := &api.Addon{}
+	err = r.Get(context.TODO(), request.NamespacedName, addon)
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			r.Log.Info("Addon deleted.")
+			err = nil
+		}
+		return
+	}
+
+	//
+	// changed.
+	err = r.changed(addon)
+	if err != nil {
+		return
+	}
+
+	// Begin staging conditions.
+	addon.Status.BeginStagingConditions()
+
+	// Ready condition.
+	if !addon.Status.HasBlockerCondition() {
+		addon.Status.SetCondition(libcnd.Condition{
+			Type:     libcnd.Ready,
+			Status:   "True",
+			Category: "Required",
+			Message:  "The addon is ready.",
+		})
+	}
+
+	// End staging conditions.
+	addon.Status.EndStagingConditions()
+
+	// Apply changes.
+	addon.Status.ObservedGeneration = addon.Generation
+	err = r.Status().Update(context.TODO(), addon)
+	if err != nil {
+		return
+	}
+
+	// Done
+	return
+}
+
+//
+// changed.
+func (r *Reconciler) changed(addon *api.Addon) (err error) {
+	if addon.Name != "admin" {
+		return
+	}
+	err = r.deleted(addon)
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	for _, mount := range addon.Spec.Mounts {
+		m := &model.Volume{}
+		m.Name = mount.Name
+		err = r.DB.Create(m).Error
+		if err != nil {
+			err = liberr.Wrap(err)
+			return
+		}
+	}
+
+	return
+}
+
+//
+// deleted.
+func (r *Reconciler) deleted(addon *api.Addon) (err error) {
+	err = r.DB.Delete(&model.Volume{}, "id > 0").Error
```

Comment:
Just curious, this query should delete all (or first of) Volumes with ID > 0?

Reply:
Yeah, it's the only way I know of to get the ORM to delete all since IDs begin with 1.